File: php/array_keys.php
<?php 

function title($v)
{
  echo "\n\n\n\n" . $v . "\n\n";
}

function observation($v)
{
  echo "\n$v";
}

$a1 = array("a", "b");
$a2 = array("a", "b"); // same value, different reference
$a3 = array("a", "c"); // different value (and different reference)
$a4 = array("b", "a"); // different order

$m1 = array();

/*
 * some dummy value is put at position 3 to act as an arbitrary offset 
 * that ensures that tests do not work just for the corner 
 * case of the needle being at first position of the haystack
 */  
$m1[3] = array("dummy1", "dummy2");
 
$m1[5] = $a1;



echo "<pre>";

title('$a1');
var_export($a1);

title('$m1');
var_export($m1);

title('$a1');
var_export($a1);

title('array_keys($a1, "b")');
$res1 = array_keys($a1, "b");
var_export($res1);
observation('vanilla array_keys');

title('array_keys($m1, $a1)');
$res2 = array_keys($m1, $a1);
var_export($res2);
observation('can find same array in matrix');

title('array_keys($m1, $a2)');
$res3 = array_keys($m1, $a2);
var_export($res3);
observation('different array with same values is evaluated as a hit. so it works by value');

title('array_keys($m1, $a3)');
$res4 = array_keys($m1, $a3);
var_export($res4);
observation('different array with different values is evaluated as a miss. of course.');

title('array_keys($m1, $a4)');
$res5 = array_keys($m1, $a4);
var_export($res5);
observation('different order of elements is evaluated as a miss.');

$a1 = array_reverse($a1);
title('after array_reverse($a1)');

title('$a1');
var_export($a1);

title('$m1');
var_export($m1);
observation('a1 is not reversed in m1. so a1 was put into m1 by value.');

title('array_keys($m1, $a1)');
$res6 = array_keys($m1, $a1);
var_export($res6);
observation('can not find reversed array in matrix, since it works by value.');



echo "</pre>";

if (!class_exists('PHPUnit_Framework_TestCase')) exit;

class Test extends PHPUnit_Framework_TestCase
{
  public function test()
  {
    global $res1;
    $this->assertEquals([0 => 1], $res1);
  }

  public function test2()
  {
    global $res2;
    $this->assertEquals([0 => 5], $res2);
  }

  public function test3()
  {
    global $res3;
    $this->assertEquals([0 => 5], $res3);
  }

  public function test4()
  {
    global $res4;
    $this->assertEquals([], $res4);
  }

  public function test5()
  {
    global $res5;
    $this->assertEquals([], $res5);
  }

  public function test6()
  {
    global $res6;
    $this->assertEquals([], $res6);
  }

}

?>

File: perl/array_keys.pl
use strict;
use warnings;
use Data::Dumper;
use Test::More;

sub title
{
  my $v = shift;
  print "\n\n\n\n$v\n\n";
}

sub var_dump
{
  my $v = shift;
  print Dumper $v;
}

sub observation
{
  my $v = shift;
  print "\n$v";
}



my @a1 = ("a", "b");
my @a2 = ("a", "b"); # same value, different reference
my @a3 = ("a", "c"); # different value (and different reference)
my @a4 = ("b", "a"); # different order
my @a5 = ();

my %m1 = ();
$m1{3} = ["dummy1", "dummy2"];
$m1{5} = \@a1;



print "Content-Type: text/html\n\n";
print "<pre>";

title("a1");
var_dump(\@a1);

title("m1");
var_dump(\%m1);

title('"b" ~~ a1');
my $res = "b" ~~ @a1;
ok($res eq 1, "smart match string in array works");

title('a1 ~~ m1');
$res = @a1 ~~ [values %m1];
ok($res eq '', "smart match array in hash does not work");

title('grep madness a1 in m1');
($res) = grep { $m1{$_} eq \@a1 } keys %m1;
ok($res eq '5', "can find same array in hash");

title('grep madness a2 in m1');
($res) = grep { $m1{$_} eq \@a2 } keys %m1;
ok(!defined $res, "the hash actually stores the reference, so a different reference will not be a match");

title('grep madness a3 in m1');
($res) = grep { $m1{$_} eq \@a3 } keys %m1;
ok(!defined $res, "different array with different values is a miss. of course.");

@a1 = reverse(@a1);
title('after reversing a1');

title('a1');
var_dump(\@a1);

title('m1');
var_dump(\%m1);
observation("value appears changed in the matrix as well. works by reference.")

title('grep madness a1 in m1');
($res) = grep { $m1{$_} eq \@a1 } keys %m1;
var_dump($res);
ok($res eq '5', "can find reversed array in hash because it works by reference");



done_testing();

print "</pre>";

